### PR TITLE
Fixes #10250 - simple SSH BMC provider

### PR DESCRIPTION
**Pull request by lzap**
Also removed nonsense test which always fail on my configuration (I have a
different TFTP root set on my dev setup).

### CODE REVIEW

**Comment by logicminds** [comment]
Can you provide test coverage for the new ssh provider?   What user will the ssh command login as?  If not the root user you will need to use sudo on the remote machine to run those commands.  This most likely will not work on windows and non linux hosts.  I am also wondering if we should give the ability to pass in a different ssh key file reference (use this key location instead of the default).  

On a side note I wonder if we should add support for fence agents which would provide much more coverage of various devices if needed. https://access.redhat.com/articles/28603

**Comment by lzap** [comment]
Rebased, added user, individual commands configuration, tests.

File: config/settings.d/bmc.yml.example [hunk 1]
```diff
@@ -5,4 +5,15 @@
...
+# - ssh - limited remote control (status, reboot, turn off)
 #:bmc_default_provider: freeipmi
+
+# default user and ssh key for ssh provider
+#:bmc_ssh_user: root
+#:bmc_ssh_key: /usr/share/foreman/.ssh/id_rsa
+
+# individual commands for actions
+#:bmc_ssh_powerstatus: "true"
+#:bmc_ssh_powercycle: "shutdown -r +1"
```
Comment (dmitri-d) [inline config/settings.d/bmc.yml.example:17]: Do all of these (`powercycle` and `poweroff` in particular) need to be settings? Can they be moved into the provider itself?

File: lib/proxy/util.rb [hunk 2]
```diff
@@ -91,6 +91,29 @@ def escape_for_shell(command)
     end
   end
 
+  def shell_command(cmd, wait = true)
```
Comment (dmitri-d) [inline lib/proxy/util.rb:94]: Could you open another PR to update puppetrun providers to use this method instead the one defined in modules/puppet_proxy_common/runner.rb?

File: test/bmc/bmc_api_ssh_test.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  include Rack::Test::Methods
+
+  def app
+    Proxy::BMC::Api.new
+  end
+
+  def setup
+    @host = "somehost"
+    @args = { :bmc_provider => "ssh" }
+    Proxy::BMC::Plugin.settings.stubs(:poweron).returns('echo poweron')
```
Comment (dmitri-d) [inline test/bmc/bmc_api_ssh_test.rb]: Please use `Proxy::BMC::Plugin.load_test_settings()` instead of stubbing (less pain down the road).

**Review by dmitri-d — CHANGES_REQUESTED** [review]

File: config/settings.d/bmc.yml.example (hunk 1 above)
Comment (lzap) [reply config/settings.d/bmc.yml.example:17]: As @logicminds stated above, there are some fencing utilities which works from shell, so you could possibly configure them here and use the ssh provider with them.

File: lib/proxy/util.rb (hunk 2 above)
Comment (lzap) [reply lib/proxy/util.rb:94]: Absolutely, can you ping me after you merge this so I can fork it easily :-)

File: test/bmc/bmc_api_ssh_test.rb (hunk 3 above)
Comment (lzap) [reply test/bmc/bmc_api_ssh_test.rb]: Added.

**Comment by dmitri-d** [comment]
[test]

**Comment by dmitri-d** [comment]
[test]

**Comment by dmitri-d** [comment]
Unhappy rubocop is unhappy.

**Comment by lzap** [comment]
Done!

**Comment by dmitri-d** [comment]
Merged as d959942. Thanks, @lzap!

**Comment by mmoll** [comment]
anybody up for a corresponding PR to puppet-foreman_proxy? :)

**Comment by lzap** [comment]
Michael Moll píše v Út 04. 10. 2016 v 12:28 -0700:

> anybody up for a corresponding PR to puppet-foreman_proxy? :)

I use this mainly for development, I don't see many use cases in the
production environment.

## 

Later,
 Lukas "lzap" Zapletal

**Pull request closed without merging**